### DISCUSSION TITLE
Multiple functions with same name

### DISCUSSION THREAD

**Discussion by asokol123** [PyO3/pyo3#4522] [Questions]
Is it possible to create multiple functions (or methods) with same name, but different signatures?
```rust
use pyo3::prelude::*;

/// Formats the sum of two numbers as string.
#[pyfunction(name = "func")]
fn sum_as_string(a: usize, b: usize) -> PyResult<String> {
    Ok((a + b).to_string())
}

#[pyfunction(name = "func")]
fn default_impl() -> PyResult<String> {
    Ok("default string".into())
}

/// A Python module implemented in Rust.
#[pymodule]
fn tmp(m: &Bound<'_, PyModule>) -> PyResult<()> {
    m.add_function(wrap_pyfunction!(sum_as_string, m)?)?;
    m.add_function(wrap_pyfunction!(test_func, m)?)?;
    Ok(())
}
```

**Comment by birkenfeld** [comment]
No - after all, Python doesn't have overloading either. PyO3 would have to define an ad-hoc way of matching parameters against signatures, and that is a minefield that I don't think we want to enter. The safe way is to do the matching yourself, in the exact way that you think it should work.
